apps/decoder_csv/lib/decoder/csv.ex

```diff
@@ -25,25 +25,25 @@ defmodule Decoder.Csv do
   defimpl Decoder, for: __MODULE__ do
     def lines_or_bytes(_t), do: :line
 
-    def decode(t, stream) do
-      stream
-      |> Stream.transform(%{skip: t.skip_first_line}, fn chunk, %{skip: skip} = acc ->
-        parsed_chunk = parse_chunk(chunk, t.headers, skip)
-        {[parsed_chunk], %{acc | skip: false}}
-      end)
+    def decode(t, messages) do
+      do_decode(messages, t.skip_first_line and !already_skipped?())
+      |> Enum.map(&parse(&1, t.headers))
     end
 
-    defp parse_chunk(chunk, headers, skip) do
-      {buffer, _} =
-        Enum.reduce(chunk, {[], skip}, fn
-          message, {buffer, false} ->
-            {[parse(message, headers) | buffer], false}
+    defp do_decode(messages, false = _should_skip?), do: messages
```

Comment:
do_decode doesn't really make sense anymore since this function no longer decodes anything.

Reply:
You're right. I'll see what changes I can make. Or maybe this last refactor took it too far?

Reply:
I think the last refactor works well, just need to update names to better represent purpose of the function.

Reply:
Yea i think do_decode is really "process_headers"

Reply:
or "ensure_the_headers_are_skipped_or_not_based_upon_what_the_user_asked_us_to_do_and_if_we_did_it_already"